File: packages/core/src/mutation.ts
import { BehaviorSubject, type Observable } from 'rxjs';
import type { MutationOptions, MutationState } from './types';
import { deriveStatusFlags, createInitialMutationState } from './types';

export class Mutation<TData = unknown, TVariables = unknown, TError = Error> {
  private state$: BehaviorSubject<MutationState<TData, TError>>;
  private options: MutationOptions<TData, TVariables, TError>;

  get state(): MutationState<TData, TError> {
    return this.state$.value;
  }

  get state$Observable(): Observable<MutationState<TData, TError>> {
    return this.state$.asObservable();
  }

  constructor(options: MutationOptions<TData, TVariables, TError>) {
    this.options = options;

    // Initialize BehaviorSubject with default state
    this.state$ = new BehaviorSubject<MutationState<TData, TError>>(
      createInitialMutationState<TData, TError>(),
    );
  }

  subscribe(
    observerOrCallback:
      | {
          next: (state: MutationState<TData, TError>) => void;
          error?: (err: unknown) => void;
          complete?: () => void;
        }
      | ((state: MutationState<TData, TError>) => void),
  ): () => void {
    const observer =
      typeof observerOrCallback === 'function'
        ? { next: observerOrCallback }
        : observerOrCallback;
    const subscription = this.state$.subscribe(observer);
    return () => subscription.unsubscribe();
  }

  private updateState(partial: Partial<MutationState<TData, TError>>): void {
    const prevState = this.state$.value;
    const newStatus = partial.status ?? prevState.status;
    const nextState = {
      ...prevState,
      ...partial,
      status: newStatus,
      ...deriveStatusFlags(newStatus),
    };
    this.state$.next(nextState);
  }

  async mutate(variables: TVariables): Promise<TData> {
    this.updateState({ status: 'loading' });

    try {
      const data = await this.options.mutationFn(variables);
      this.updateState({
        status: 'success',
        data,
        error: null,
      });

      this.options.onSuccess?.(data, variables);
      this.options.onSettled?.(data, null, variables);

      return data;
    } catch (error) {
      const err = error as TError;
      this.updateState({
        status: 'error',
        error: err,
      });

      this.options.onError?.(err, variables);
      this.options.onSettled?.(undefined, err, variables);

      throw err;
    }
  }

  reset(): void {
    this.state$.next(createInitialMutationState<TData, TError>());
  }
}

File: packages/py-core/ts_query/mutation.py
"""Port of packages/core/src/mutation.ts."""

from __future__ import annotations

from typing import Any, Callable, Dict, Union

from .subject import BehaviorSubject, Observer, Subscription
from .types import (
    UNDEFINED,
    MutationOptions,
    MutationState,
    _merge_state,
    create_initial_mutation_state,
)


class Mutation:
    def __init__(self, options: MutationOptions) -> None:
        self._options = options
        self._state = BehaviorSubject(create_initial_mutation_state())

    @property
    def state(self) -> MutationState:
        return self._state.value

    @property
    def state_observable(self) -> BehaviorSubject:
        return self._state

    def subscribe(
        self, observer_or_callback: Union[Observer, Callable[[MutationState], None], Dict[str, Any]]
    ) -> Callable[[], None]:
        subscription: Subscription = self._state.subscribe(observer_or_callback)
        return subscription.unsubscribe

    def _update_state(self, partial: Dict[str, Any]) -> None:
        self._state.next(_merge_state(self._state.value, partial))

    async def mutate(self, variables: Any = None) -> Any:
        self._update_state({"status": "loading"})

        try:
            data = await self._options.mutation_fn(variables)
        except BaseException as err:  # noqa: BLE001 - re-raised below
            self._update_state({"status": "error", "error": err})

            if self._options.on_error:
                self._options.on_error(err, variables)
            if self._options.on_settled:
                self._options.on_settled(UNDEFINED, err, variables)

            raise

        self._update_state({"status": "success", "data": data, "error": None})

        if self._options.on_success:
            self._options.on_success(data, variables)
        if self._options.on_settled:
            self._options.on_settled(data, None, variables)

        return data

    def reset(self) -> None:
        self._state.next(create_initial_mutation_state())
